### PR TITLE
Fix resource allocation for tasks with dependencies

### LINKED ISSUE #6663: SerializeableLock fails when used with resources
For some reason the lock appears to be locked elsewhere.  This only happens when I have two threads and two processes, restricting usage with resources:

```python
from dask.distributed import Client
import time
from dask.utils import SerializableLock


def dummy(i: int) -> int:
    return i


def constrained(lock: SerializableLock, dummy: int):
    assert not lock.locked()
    with lock:
        time.sleep(0.005)
    assert not lock.locked()


N_TASKS = 1000
lock = SerializableLock()
futures = []

if __name__ == "__main__":
    with Client(resources={"foo": 1}, n_workers=2, threads_per_worker=2) as client:
        for i in range(N_TASKS):
            x = client.submit(dummy, i)
            futures.append(client.submit(constrained, lock, x, resources={"foo": 1}))
        client.gather(futures)
```

Things that might be wrong:

1. Resources are bad?  Should check to see what other tasks are running when this failure occurs
2. Accidentally shared state between the two processes?  Should try running this maybe with `dask-worker` processes created separately

### PR DESCRIPTION
**Pull request by hendrikmakait**
Closes #6663 
Blocked by 
* ~#6672~
* ~#6687~
* ~#6690~
* ~#6697~

- [ ] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -1888,6 +1888,9 @@ def _transition_from_resumed(
             assert next_state in {"waiting", "fetch"}, next_state
             assert ts._previous in {"executing", "flight"}, ts._previous
 
+            if ts._previous == "executing":
```
Comment (hendrikmakait) [inline distributed/worker_state_machine.py:1891]: TODO: Write test that ensures that we do in fact release IFF the task is finished.

**Comment by crusaderky** [comment]
Please merge from main. Also there are failing tests

File: distributed/tests/test_worker_state_machine.py [hunk 2]
```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
...
+            nbytes=8,
+            type=None,
+            stimulus_id="success",
+        ),
+    )
+    assert ws.tasks["x"].state == "memory"
+    assert ws.available_resources == {"R": 1}
+
+
+@pytest.mark.xfail(reason="distributed#6565")
```
Comment (hendrikmakait) [inline distributed/tests/test_worker_state_machine.py:1039]: `XFAIL` to be tackled in a follow-up PR. #6565 is already tracking this issue.

File: distributed/tests/test_worker_state_machine.py [hunk 3]
```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
...
+            nbytes=8,
+            type=None,
+            stimulus_id="s2",
+        ),
+    )
+    assert ws.tasks["x"].state == "memory"
+    assert ws.available_resources == {"R": 1}
+
+
+@pytest.mark.xfail(reason="distributed#6682")
```
Comment (hendrikmakait) [inline distributed/tests/test_worker_state_machine.py:1162]: `XFAIL` to be tackled in a follow-up PR. #6682 has been created to tackle this issue.

**Review by hendrikmakait — COMMENTED** [review]
Rebased onto the latest master to reduce the diff. Two tests have been `XFAIL`ed to be tackled in independent PRs. The corresponding issues are #6565 and #6682.

**Comment by hendrikmakait** [comment]
Failing tests are known flakes.

File: distributed/worker_state_machine.py (hunk 1 above)
Comment (crusaderky) [reply distributed/worker_state_machine.py:1890]: This and line 1888 are missing the use case of 'long-running'. I'll open a follow-up PR about it, no point adding scope creep here.

I'd like to have the test you mention within this PR please.

File: distributed/worker_state_machine.py [hunk 4]
```diff
@@ -3116,6 +3120,21 @@ def validate_state(self) -> None:
...
+            for ts in self.tasks.values():
+                if not ts.resource_restrictions:
+                    continue
+                if ts.state in {"executing", "long-running", "cancelled"} or (
+                    ts.state == "resumed"
+                    and ts._previous in {"executing", "long-running"}
+                ):
+                    allocated += ts.resource_restrictions.get(resource, 0)
+            assert available + allocated == total
+
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:3137]:
```suggestion
        self._validate_resources()
    
    @property
    def all_running_tasks(self) -> set[TaskState]:
        """All tasks that are currently running.
        These are:
        
        - ``ts.status`` == ``executing`, ``long-running``, or ``cancelled``
        - ``ts.status` == ``resumed`` and ``ts._previous`` == ``executing`` or ``long-running``
        """
        # Note: tasks in "cancelled" and "resumed" state are still in either of these sets
        return self.executing | {self.tasks[key] for key in self.long_running}

    def _validate_resources(self) -> None:
        """Assert that available_resources + resources held by tasks = total_resources"""
        total = self.total_resources.copy()
        for k, v in self.available_resources.items():
            total[k] -= v
        for ts in self.all_running_tasks:
            for k, v in ts.requires_resources.items():
                total[k] -= v

         assert all(abs(v) < 1e-9) for v in total.values())
```

Please add unit tests for `all_running_tasks` that cover the cancelled and resumed use cases

File: distributed/tests/test_worker_state_machine.py [hunk 5]
```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
...
+        ws.story("x"),
+        [
+            ("x", "compute-task", "released"),
+            ("x", "released", "waiting", "waiting", {"x": "ready"}),
+            ("x", "waiting", "ready", "waiting", {"x": "constrained"}),
+            ("x", "waiting", "constrained", "constrained", {"x": "executing"}),
+            ("x", "constrained", "executing", "executing", {}),
+        ],
+        strict=False,
+    )
```
Comment (crusaderky) [inline distributed/tests/test_worker_state_machine.py:1022]: As a general design note, we'd rather avoid using WorkerState.story as it is a level of abstraction that is well below that of events.
```suggestion
```

File: distributed/tests/test_worker_state_machine.py [hunk 6]
```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
...
+            stimulus_id="compute",
+        )
+    )
+    assert ws.tasks["x"].state == "flight"
+    assert ws.tasks["y"].state == "waiting"
+    assert ws.available_resources == {"R": 1}
+
+    ws.handle_stimulus(
+        GatherDepSuccessEvent("gather-dep-done", "127.0.0.1:1235", 8, {"x": 1.0})
+    )
```
Comment (crusaderky) [inline distributed/tests/test_worker_state_machine.py:1092]: This test has little benefit beyond this line. You should just test that this event return a Execute instruction, that the task state is 'executing', and that available_resources have been decreased.

File: distributed/tests/test_worker_state_machine.py [hunk 7]
```diff
@@ -1002,6 +1005,195 @@ def test_gather_priority(ws):
...
+
+    instructions = ws.handle_stimulus(
+        ComputeTaskEvent.dummy(
+            key="y",
+            who_has={"x": ["127.0.0.1:1235"]},
+            nbytes={"x": 8},
+            stimulus_id="compute",
+        )
+    )
+    assert ws.tasks["x"].state == "resumed"
```
Comment (crusaderky) [inline distributed/tests/test_worker_state_machine.py:1178]: Could you review this?
From reading the code (see `_transition_from_resumed`), "resumed" should be *exclusively* on one of the following loops:

- executing or long-running -> cancelled -> fetch
- flight -> cancelled -> waiting

in the executing -> cancelled -> waiting loop that you implemented here, I expect ts.state to be 'executing'.

**Review by crusaderky — CHANGES_REQUESTED** [review]
.

File: distributed/tests/test_worker_state_machine.py (hunk 5 above)
Comment (hendrikmakait) [reply distributed/tests/test_worker_state_machine.py:1022]: Sounds good, I wanted to ensure that we go through the "constrained" state in all cases, but I guess testing for resource allocation is the only thing we're trying to achieve in the end.

File: distributed/worker_state_machine.py (hunk 1 above)
Comment (hendrikmakait) [reply distributed/worker_state_machine.py:1890]: `ts._previous` should never happen to begin with in this method. I'll clean up the code in here in a separate PR.

File: distributed/tests/test_worker_state_machine.py (hunk 6 above)
Comment (hendrikmakait) [reply distributed/tests/test_worker_state_machine.py:1098]: Fair point.

File: distributed/utils_test.py [hunk 8]
```diff
@@ -2462,7 +2462,6 @@ def ws_with_running_task(ws, request):
             SecedeEvent(key="x", compute_duration=1.0, stimulus_id="secede")
         )
     assert ws.tasks["x"].state == request.param
-    assert ws.available_resources == {"R": 0}
```
Comment (hendrikmakait) [inline distributed/utils_test.py:2465]: Moved this into the dedicated `test_ws_with_running_task`

File: distributed/tests/test_worker_state_machine.py [hunk 9]
```diff
@@ -1033,6 +1036,98 @@ def test_gather_priority(ws):
     ]
 
 
+@pytest.mark.parametrize("state", ["executing", "long-running"])
+def test_running_constrained_task_acquires_resources(state, ws):
```
Comment (hendrikmakait) [inline distributed/tests/test_worker_state_machine.py:1040]: Duplicating logic from `ws_with_running_task` and `test_ws_with_running_task` to have an explicit test focused on resource restrictions that is resilient to changes to those functions.

File: distributed/tests/test_worker_state_machine.py (hunk 7 above)
Comment (hendrikmakait) [reply distributed/tests/test_worker_state_machine.py]: This test does not create an `executing -> cancelled -> waiting` loop, but an `executing -> cancelled -> fetch` loop by cancelling `x` and then gathering it as a dependency to `y`.

**Comment by hendrikmakait** [comment]
CI issues are being caused by #6692

File: distributed/tests/test_worker_state_machine.py (hunk 2 above)
Comment (crusaderky) [reply distributed/tests/test_worker_state_machine.py:1077]: This is already green

File: distributed/tests/test_worker_state_machine.py [hunk 10]
```diff
@@ -1053,6 +1056,98 @@ def test_gather_priority(ws):
...
+            resource_restrictions={"R": 1},
+            stimulus_id="compute",
+        )
+    )
+    assert ws.tasks["x"].state == "flight"
+    assert ws.tasks["y"].state == "waiting"
+    assert ws.available_resources == {"R": 1}
+
+    instructions = ws.handle_stimulus(
+        GatherDepSuccessEvent("gather-dep-done", "127.0.0.1:1235", 8, {"x": 1.0})
```
Comment (crusaderky) [inline distributed/tests/test_worker_state_machine.py:1110]: This is unreadable, please never go full-positional

File: distributed/worker_state_machine.py [hunk 11]
```diff
@@ -1963,6 +1964,8 @@ def _transition_from_resumed(
 
             if ts._previous in ("executing", "long-running"):
                 self._release_resources(ts)
+                self.executing.discard(ts)
+                self.long_running.discard(ts)
```
Comment (crusaderky) [inline distributed/worker_state_machine.py:1968]: This is actually superfluous as we're invoking purge_state afterwards. But I think it's a good idea to have it for cleanliness' sake.

**Comment by crusaderky** [comment]
See code review: https://github.com/dask/distributed/pull/6676/commits/74defe412bdb22fda15d41ff1ca29f8ea9fda8aa
There's an xfail in `test_resumed_task_releases_resources`:
```
E       AssertionError: assert {'R': 0} == {'R': 1}
E         Differing items:
E         {'R': 0} != {'R': 1}
E         Full diff:
E         - {'R': 1}
E         ?       ^
E         + {'R': 0}
E         ?       ^
```
The culprits are, again, #6689 + #6693.
`_transition_from_resumed` is not hit in case of error, so the change in this PR does not have the effect you hoped.

I'm going to merge the PR as-is, but it means it doesn't close #6663.
I'll resolve with #6689, #6693, and #6663 in a single PR before I leave.

**Review by crusaderky — APPROVED** [review]

**Comment by hendrikmakait** [comment]
There is a separate issue open around the issue with failing resumed tasks: #6682. This PR should solve the reproducer in #6663 (it works just fine on my machine), so I'd close that issue with merging and keep #6682 open.

**Comment by hendrikmakait** [comment]
@crusaderky: Thanks for the thorough review!

File: distributed/tests/test_worker_state_machine.py (hunk 2 above)
Comment (hendrikmakait) [reply distributed/tests/test_worker_state_machine.py]: Good catch, somehow missed that.

**Pull request merged**